### ISSUE TITLE
Bug: extends does not allow user-specified `files` to win / `extends` will override user-specified files in a way that does not let the user merge in their globs / the config-inspector does not correctly show the reality of extends (but it does show what I want to be reality)

### ISSUE THREAD

**Issue by NullVoxPopuli** [eslint/eslint#20351]
### Environment

Node version: n/a
npm version: n/a
Local ESLint version: 9.x
Global ESLint version: n/a
Operating System: n/a

### What parser are you using?

@typescript-eslint/parser

### What did you do?

There is an issue here about this behavior:

https://github.com/typescript-eslint/typescript-eslint/issues/11774

where, it seems that the typescript-eslint project is specifying something like this in their configs:
- https://github.com/typescript-eslint/typescript-eslint/issues/11774#issuecomment-3567061085
  ```js
  // recommendedTypeChecked
  [
    { /* ... */ },
    {
       files: [ '**/*.ts', '**/*.tsx', '**/*.mts', '**/*.cts' ],
       /* ... */ 
    }
    /* ... */,
  ]
  ```

and the [recommendation](https://github.com/typescript-eslint/typescript-eslint/issues/11774#issuecomment-3567087266) by @bradzacher was to do:
```js
{
  ...tseslint.configs.eslintRecommended,
  files: [yourGlobsHere],
}
```
but it doesn't work because [`extends` does in intersection instead of a union](https://github.com/typescript-eslint/typescript-eslint/issues/11774#issuecomment-3567095462)

So what this has lead us to do is something like this:
```js
 {
   files: ['**/*.{ts,gts}'],
   extends: [
      ...ts.configs.recommendedTypeChecked,
      {
        ...ts.configs.eslintRecommended,
        files: undefined,
      },
    ],
  },
```

which... is super gross, as I feel like this should work:
```js
 {
   files: ['**/*.{ts,gts}'],
   extends: [
      ...ts.configs.recommendedTypeChecked,
    ],
  },
```
and by work, I mean, the total config for this config object should only apply to `**/*.{ts,gts}`, and ignore whatever other files are in the `extends` -- as I want to override.

### What did you expect to happen?

user-provided `files` wins.

### What actually happened?

no `files` entry is used, an intersection is used, which is `[]`

### Link to Minimal Reproducible Example

<details><summary>previous repro invalid (see comment below)</summary>

https://stackblitz.com/edit/ember-cli-editor-output-ckayeg5x?file=package.json

actually... this config inspector seems to say that what I want is what is already happening

<img width="1016" height="1024" alt="Image" src="https://github.com/user-attachments/assets/e96b99cc-3863-486d-9f74-92dc219efdfb" />

I need to conspire with @bendemboski real quick

</details>

### Participation

- [ ] I am willing to submit a pull request for this issue.

**Comment by NullVoxPopuli** [comment]
Here is an actual minimal repro: https://stackblitz.com/edit/vitejs-vite-c6vkv4ft?file=eslint.config.mjs

Turns out typescript-eslint's ts.config was fixing this for me -- which even more leads me to believe that eslint has the wrong behavior here

**Comment by NullVoxPopuli** [comment]
Here is a modified version of that minimal repro which shows a linting error:
- https://stackblitz.com/edit/vitejs-vite-51gr37xf?file=eslint.config.mjs,package.json,subdir%2Ffile.js
(unexpected, because the eslint config disables it)

_BUT_ it seems the inspector is also wrong, because the config inspector thinks the lint rule should be disabled


Unexpected:
<img width="654" height="266" alt="Image" src="https://github.com/user-attachments/assets/4d9f312b-554b-4057-882d-20223c6e7924" />

from config:
```js
import { defineConfig } from 'eslint/config';
import js from '@eslint/js';

export default defineConfig(js.configs.recommended, {
  files: ['root.js'],
  extends: [
    {
      files: ['subdir/*.js'],
      rules: { 'no-unused-vars': 'off' },
    },
  ],
});
```

The inspector reports what I would expect, but CLI behavior does not agree
<img width="602" height="411" alt="Image" src="https://github.com/user-attachments/assets/f215c535-eefe-4dcd-a1d8-fd7081cc8556" />

**Comment by mdjermanovic** [comment]
> `extends` does in intersection

This works as intended per the RFC:

https://github.com/eslint/rfcs/tree/main/designs/2024-config-extends#base-object-with-files-and-ignores-and-extended-object-with-files-or-ignores


> So what this has lead us to do is something like this:
>
> ```js 
>  {
>    files: ['**/*.{ts,gts}'],
>    extends: [
>       ...ts.configs.recommendedTypeChecked,
>       {
>         ...ts.configs.eslintRecommended,
>         files: undefined,
>       },
>     ],
>   },
> ```

You can use `ignores` to exclude those file extensions specified in extended config's `files` that you don't want the configuration to apply to:

https://eslint.org/docs/latest/use/configure/configuration-files#excluding-files-with-ignores

> _BUT_ it seems the inspector is also wrong, because the config inspector thinks the lint rule should be disabled

It could be that config inspector doesn't handle intersections (arrays in the `files` array) correctly. Can you please open an issue in the config-inspector repository?

~https://github.com/eslint/code-explorer~ https://github.com/eslint/config-inspector

**Comment by NullVoxPopuli** [comment]
Ye, Here ya go 

https://github.com/eslint/config-inspector/issues/190

**Comment by NullVoxPopuli** [comment]
> works as intended 

```js
import { defineConfig } from 'eslint/config';
import js from '@eslint/js';

export default defineConfig(js.configs.recommended, {
  files: ['root.js'],
  extends: [
    {
      files: ['subdir/*.js'],
      rules: { 'no-unused-vars': 'off' },
    },
  ],
});
```

So this not matching root.js is intended?

Why wouldn't you union them like the inspector says?

**Comment by nzakas** [comment]
Note: The correct repo is https://github.com/eslint/config-inspector.

**Comment by mdjermanovic** [comment]
> import { defineConfig } from 'eslint/config';
> import js from '@eslint/js';
> 
> export default defineConfig(js.configs.recommended, {
>   files: ['root.js'],
>   extends: [
>     {
>       files: ['subdir/*.js'],
>       rules: { 'no-unused-vars': 'off' },
>     },
>   ],
> });
> 
>     
>   
> 
> So this not matching root.js is intended?

Yes, because in this case paths should match both `root.js` and `subdir/*.js` patterns, but `root.js` path doesn't match `subdir/*.js` pattern.

> Why wouldn't you union them like the inspector says?

This was discussed on the RFC: https://github.com/eslint/rfcs/pull/126#discussion_r1843063505

**Comment by NullVoxPopuli** [comment]
Can we have a util that changes that? this feels bad

(especially w/ this gaslighting: https://github.com/eslint/config-inspector/issues/190)
